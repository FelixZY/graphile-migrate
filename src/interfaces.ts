/** Represents the graphile_migrate.current type in the DB */
export interface DbCurrent {
  filename: string;
  content: string;
  date: Date;
}
