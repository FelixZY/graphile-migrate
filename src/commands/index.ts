export { commit } from "./commit";
export { compile } from "./compile";
export { init } from "./init";
export { migrate } from "./migrate";
export { reset } from "./reset";
export { run } from "./run";
export { status } from "./status";
export { uncommit } from "./uncommit";
export { watch } from "./watch";
