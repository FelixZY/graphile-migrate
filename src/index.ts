export * from "./commands";
export { defaultLogger } from "./logger";
export { Settings } from "./settings";
export { LogFunctionFactory, Logger, LogLevel } from "@graphile/logger";
