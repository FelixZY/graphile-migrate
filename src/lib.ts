export function mergeWithoutClobbering(
  original: { [key: string]: string | undefined },
  newStuff: { [key: string]: string | undefined },
  message: string,
): { [key: string]: string | undefined } {
  const result = { ...original };
  for (const key in newStuff) {
    if (typeof result[key] === "undefined" || result[key] === newStuff[key]) {
      result[key] = newStuff[key];
    } else {
      throw new Error(
        `Refusing to clobber '${key}' (from '${original[key]}' to '${newStuff[key]}'): ${message}`,
      );
    }
  }

  return result;
}

export function isLoggedError(error: unknown): boolean {
  return (
    typeof error === "object" &&
    error !== null &&
    "_gmlogged" in error &&
    error._gmlogged === true
  );
}

export function errorCode(e: unknown): string | null {
  return typeof e === "object" &&
    e !== null &&
    "code" in e &&
    typeof e.code === "string"
    ? e.code
    : null;
}
