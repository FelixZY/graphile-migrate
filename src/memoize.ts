/* eslint-disable @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-assignment */
export default function memoize<T extends (...args: Array<any>) => any>(
  fn: T,
): (...funcArgs: Parameters<T>) => ReturnType<T> {
  let lastArgs: Parameters<T>;
  let lastResult: ReturnType<T>;
  return (...args: Parameters<T>): ReturnType<T> => {
    if (
      lastArgs &&
      args.length === lastArgs.length &&
      args.every((arg, i) => arg === lastArgs[i])
    ) {
      return lastResult;
    } else {
      lastArgs = args;
      lastResult = fn(...args);
      return lastResult;
    }
  };
}
