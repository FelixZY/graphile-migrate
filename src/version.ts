// This file is autogenerated by /scripts/version.mjs
export const version = "2.0.0-rc.1";
