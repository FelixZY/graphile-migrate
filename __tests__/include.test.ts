import "./helpers";

import * as mockFs from "mock-fs";

import { compileIncludes } from "../src/migration";
import { ParsedSettings, parseSettings } from "../src/settings";

let old: string | undefined;
let settings: ParsedSettings;
beforeAll(async () => {
  old = process.env.DATABASE_AUTHENTICATOR;
  process.env.DATABASE_AUTHENTICATOR = "dbauth";
  settings = await parseSettings({
    connectionString: "postgres://dbowner:dbpassword@dbhost:1221/dbname",
    placeholders: {
      ":DATABASE_AUTHENTICATOR": "!ENV",
    },
    migrationsFolder: 'migrations'
  });
});
afterAll(() => {
  process.env.DATABASE_AUTHENTICATOR = old;
});

afterEach(() => {
  mockFs.restore();
});


it("compiles an included file", async () => {
  mockFs({
    "migrations/fixtures/foo.sql": "select * from foo;",
  });
  expect(
    await compileIncludes(
      settings,
      `\
--!include foo.sql
`,
    ),
  ).toEqual(`\
select * from foo;
`);
});

it("compiles multiple included files", async () => {
  mockFs({
    "migrations/fixtures/dir1/foo.sql": "select * from foo;",
    "migrations/fixtures/dir2/bar.sql": "select * from bar;",
    "migrations/fixtures/dir3/baz.sql": "--!include dir4/qux.sql",
    "migrations/fixtures/dir4/qux.sql": "select * from qux;",
  });
  expect(
    await compileIncludes(
      settings,
      `\
--!include dir1/foo.sql
--!include dir2/bar.sql
--!include dir3/baz.sql
`,
    ),
  ).toEqual(`\
select * from foo;
select * from bar;
select * from qux;
`);
});

it("compiles an included file, and won't get stuck in an infinite include loop", async () => {
  mockFs({
    "migrations/fixtures/foo.sql": "select * from foo;--!include foo.sql",
  });
  expect(
    await compileIncludes(
      settings,
      `\
--!include foo.sql
`,
    ),
  ).toEqual(`\
select * from foo;
`);
});

/*it("disallows calling files outside of the migrations/fixtures folder", async () => {
  mockFs({
    "migrations/fixtures/bar.sql": "",
    "outsideFolder/foo.sql": "select * from foo;",
  });
  expect(() => {
    compileIncludes(
      settings,
      `\
--!include ../../outsideFolder/foo.sql
`,
    )},
  ).toThrow();
});*/