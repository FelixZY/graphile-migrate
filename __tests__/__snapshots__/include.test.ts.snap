// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compiles an included file, and won't get stuck in an infinite include loop 1`] = `
"Circular include detected - '~/migrations/fixtures/foo.sql' is included again! Trace:
  ~/migrations/fixtures/foo.sql
  ~/migrations/current.sql"
`;

exports[`disallows calling files outside of the migrations/fixtures folder 1`] = `"Forbidden: cannot include path '~/outsideFolder/foo.sql' because it's not inside 'migrations/fixtures/'"`;
